### PR TITLE
feat: analytical frequency/hessian/eigenmatrix gradients via eigenvalue sensitivity

### LINKED ISSUE #221: feat: Analytical frequency/hessian/eigenmatrix gradients via JAX eigenvalue sensitivity
## Summary

Add analytical (autodiff) gradient support for the frequency, hessian_element, and eigenmatrix evaluators when using the JAX backend. Currently only the energy evaluator supports analytical gradients — all others fall back to finite-difference approximation (2N+1 evaluations per gradient step).

## Motivation

For a force field with P parameters and N atoms, the finite-difference gradient of a frequency-based objective requires **2P full Hessian evaluations per optimizer step** — each Hessian itself being an O(N²) autodiff computation. For the Rh-enamide system (52 atoms, ~20 parameters), that's 40 Hessian evaluations just for frequencies. Analytical gradients reduce this to **one Jacobian computation** via `jax.jacrev`, giving exact derivatives with no step-size tuning.

This is the single highest-impact optimization for mixed energy+frequency fitting, which is the most common q2mm workflow.

### Why not differentiate through `eigh` directly?

JAX's `jnp.linalg.eigh` supports reverse-mode AD, but its gradient is **undefined for degenerate eigenvalues** — the eigenvectors become non-unique within degenerate subspaces, making the backward pass unstable. Real molecules routinely have degenerate modes (E symmetry in C₃ᵥ, T symmetry in Tₐ, etc.), plus the 6 near-zero translational/rotational modes.

### Preferred approach: Manual eigenvalue sensitivity formula

The classical eigenvalue perturbation result avoids differentiating through `eigh` entirely:

```
dλ_k/dp = v_k^T · (dA/dp) · v_k
```

where `A` is the mass-weighted Hessian, `λ_k` and `v_k` are its eigenvalues/eigenvectors, and `dA/dp` is the Jacobian of the mass-weighted Hessian w.r.t. force field parameters. This formula is well-defined even for degenerate eigenvalues (the sum of eigenvalue derivatives within a degenerate subspace is always correct).

The full chain for frequencies is:

```
params → E(params, coords)                          # energy function (JAX)
      → d²E/dcoords² = H(params)                    # Hessian via jax.hessian
      → dH/dp via jax.jacrev(jax.hessian(...))       # 3rd-order mixed derivative
      → mass-weight dH/dp                            # linear, trivially differentiable
      → dλ_k/dp = v_k^T · mw(dH/dp) · v_k           # eigenvalue sensitivity
      → d(freq_k)/dp = d(freq)/dλ · dλ_k/dp          # chain rule through sqrt/unit conversion
```

### Architecture

**Engine provides one new capability:**
- `supports_analytical_hessian_gradients()` → bool (base returns False; JaxEngine returns True)
- `hessian_and_param_jacobian(mol, ff)` → `(H, dH/dp)` where H is (3N, 3N) and dH/dp is (3N, 3N, n_params)

This is computed as:
```python
jax.jacrev(jax.hessian(energy_fn, argnums=coords), argnums=params)
```

**Each evaluator chains through its own observable:**

| Evaluator | Formula | Notes |
|-----------|---------|-------|
| **FrequencyEvaluator** | `dλ_k/dp = v_k^T · mw(dH/dp) · v_k`, then chain through `√λ → cm⁻¹` | Eigenvalue sensitivity formula |
| **HessianElementEvaluator** | `d(H[i,j])/dp = dH_dp[i, j, :]` | Direct extraction from Jacobian |
| **EigenmatrixEvaluator** | `d(Q^T H Q)/dp = Q^T · (dH/dp) · Q` | QM eigenvectors Q are constant |

Then each evaluator chains through the residual formula:
```
d(score)/dp = -2 · Σᵢ [wᵢ² · (refᵢ - calcᵢ) · d(calcᵢ)/dp]
```

### Why the Hessian Jacobian is the right abstraction

1. **Computed once, used by three evaluators** — frequency, hessian_element, and eigenmatrix all derive from the Hessian
2. **The 3rd-order derivative is unavoidable** — `dH/dp = d³E/(dx² dp)` regardless of approach
3. **Avoids JAX `eigh` backward pass entirely** — eigenvalue sensitivity uses eigenvectors only in the forward direction
4. **Matches the existing pattern** — `energy_and_param_grad()` provides raw `dE/dp`, and the EnergyEvaluator chains through the residual formula

### Edge Cases

| Edge case | Impact | Mitigation |
|-----------|--------|------------|
| **Degenerate eigenvalues** | Eigenvalue sensitivity formula still works (sum over degenerate subspace is well-defined). Individual eigenvector assignment within the subspace is arbitrary but the total eigenvalue derivative is correct. | None needed — the formula handles it. |
| **Near-zero eigenvalues** (6 trivial modes) | `d(freq)/dλ = factor / (2·√\|λ\|)` diverges as `λ → 0` | Regularize: `max(\|λ\|, ε)` in the denominator. These modes are never used as fitting targets anyway. |
| **JIT compilation cost** | 3rd-order derivative trace may be slow for large molecules on first call | Cache compiled function on `JaxHandle` (same pattern as `_coord_hess_fn`) |
| **Numerical precision** | 3rd-order derivatives accumulate more floating-point error | Validate against finite-difference Jacobian in tests |

### Files to modify

1. **`q2mm/backends/base.py`** — Add `supports_analytical_hessian_gradients()` (returns False) and `hessian_and_param_jacobian()` (raises NotImplementedError)
2. **`q2mm/backends/mm/jax_engine.py`** — Implement both methods. Add `_hess_param_jac_fn` to `JaxHandle` dataclass. Use `jax.jit(jax.jacrev(jax.hessian(energy_fn, argnums=0), argnums=1))`
3. **`q2mm/models/hessian.py`** — Add `frequency_param_jacobian(H, dH_dp, symbols)` that returns `(freqs, d_freq_d_params)` using the eigenvalue sensitivity formula
4. **`q2mm/optimizers/evaluators/frequency.py`** — `supports_analytical_gradient()` checks `engine.supports_analytical_hessian_gradients()`. Implement `gradient()` using `hessian_and_param_jacobian` + `frequency_param_jacobian`
5. **`q2mm/optimizers/evaluators/hessian_element.py`** — Same pattern. `gradient()` extracts `dH[i,j]/dp` directly from Jacobian tensor.
6. **`q2mm/optimizers/evaluators/eigenmatrix.py`** — Same pattern. `gradient()` computes `Q^T (dH/dp) Q` where Q are the cached QM eigenvectors.
7. **Tests** — Validate analytical gradients against finite-difference for all three evaluators. Both pure-NumPy tests (synthetic Hessian inputs for `frequency_param_jacobian`) and JAX integration tests (end-to-end with `@pytest.mark.jax`).

### Testing strategy

The gold standard: compute analytical gradient, compare to central finite-difference gradient. For each evaluator:
1. Create a small test system (e.g., 3-atom molecule with known force field)
2. Compute analytical gradient via the new methods
3. Compute FD gradient by perturbing each parameter ±h and re-evaluating
4. Assert `np.allclose(analytical, fd, rtol=1e-4)`

## Relationship to existing issues

- **#152 (End-to-end differentiable FF fitting)** — This implements Phase 3 of that roadmap (differentiable frequency/eigenmatrix objective). Closing this makes #152 substantially closer to completion.
- **#176 (JAX-ReaxFF pattern)** — The `hessian_and_param_jacobian` method is a prerequisite for the vmapped architecture proposed there.

## Prior art

- [DMFF](https://github.com/deepmodeling/DMFF) — JAX-based differentiable FF platform, uses `jax.hessian` + `jnp.linalg.eigh` for vibrational analysis
- [JAX-ReaxFF](https://github.com/cagrikymk/JAX-ReaxFF) — Gradient-based ReaxFF optimization via JAX autodiff
- Eigenvalue sensitivity formula: standard result in structural dynamics (see e.g., Fox & Kapoor 1968, Nelson 1976)

### PR DESCRIPTION
**Pull request by ericchansen**
## Summary

Implement analytical gradient computation for three evaluator types that previously required finite-difference gradients. Uses the classical eigenvalue sensitivity formula to avoid differentiating through `eigh` (which is undefined for degenerate eigenvalues).

**Key formula:** `dλ_k/dp = v_k^T · (dA/dp) · v_k`

where `A` is the mass-weighted Hessian, `v_k` is the k-th eigenvector, and `dA/dp` is computed by JAX autodiff.

### Engine API (`base.py`, `jax_engine.py`)
- `supports_analytical_hessian_gradients()` — capability query
- `hessian_and_param_jacobian(mol, ff) → (H, dH_dp)` — returns Hessian + parameter Jacobian
- JaxEngine implementation uses `jax.jacrev(jax.hessian(E))` with JIT caching

### Three Evaluator Gradients
| Evaluator | Gradient Method |
|-----------|----------------|
| **Frequency** | Mass-weight `dH/dp`, eigenvalue sensitivity, chain through `√λ → cm⁻¹` |
| **HessianElement** | Direct `dH[i,j]/dp` extraction from Jacobian tensor |
| **Eigenmatrix** | `Q^T · (dH/dp) · Q` via einsum (QM eigenvectors `Q` cached) |

### Math Utilities (`hessian.py`)
- `frequency_param_jacobian()` — pure NumPy eigenvalue sensitivity + frequency conversion chain rule
- Near-zero eigenvalue regularization for trivial translation/rotation modes

### Protocol Update
- Added `mol_idx: int = 0` to all evaluator `gradient()` signatures for eigenmatrix QM eigenvector caching

## Tests
- **21 new tests** in `test_hessian_grad.py` covering:
  - `frequency_param_jacobian` (identity, diagonal, off-diagonal, sign-flip, zero-mass, multi-param)
  - Frequency evaluator gradient (FD validation, chain rule, unsupported engine)
  - HessianElement evaluator gradient (direct extraction, multi-param, upper-triangle)
  - Eigenmatrix evaluator gradient (Q rotation, caching, multi-param)
  - Base engine API defaults
- Updated existing tests for new analytical gradient capabilities

## Validation
- All 799 core tests pass (no backend dependencies)
- Lint and format clean
- FD validation of analytical gradients at rtol=1e-4 to 1e-5

Closes #221

### CODE REVIEW

File: q2mm/backends/mm/jax_engine.py [hunk 1]
```diff
@@ -858,6 +859,53 @@ def _energy_of_flat_coords(flat_coords: jnp.ndarray, params_: jnp.ndarray) -> jn
...
+                dh_dp = jax.jacrev(hess_fn, argnums=1)(flat_coords_, params_)
+                return h, dh_dp
+
+            handle._hess_param_jac_fn = jax.jit(_hess_and_jac)
+
+        hess_kcal_a2, dhess_dp_kcal_a2 = handle._hess_param_jac_fn(flat_coords, params)
+        return (
+            np.asarray(hess_kcal_a2) * KCALMOLA2_TO_HESSIAN_AU,
+            np.asarray(dhess_dp_kcal_a2) * KCALMOLA2_TO_HESSIAN_AU,
+        )
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/backends/mm/jax_engine.py:907]: This is not a bug. I verified empirically that `jax.jacrev(hess_fn, argnums=1)` returns shape `(3N, 3N, n_params)` — the output dimensions of the differentiated function come first, followed by the parameter axis. This already matches our convention. All 34 integration tests pass with the current axis order, including FD parity checks that would fail immediately if the axes were transposed.

File: q2mm/optimizers/evaluators/eigenmatrix.py [hunk 2]
```diff
@@ -161,17 +161,62 @@ def gradient(
...
+        grad = np.zeros(n_params)
+        for ref in references:
+            if ref.kind == "eig_diagonal":
+                idx = ref.data_idx
+                calc_value = float((qm_evecs.T @ hess @ qm_evecs)[idx, idx])
+                diff = ref.value - calc_value
+                grad += -2.0 * ref.weight**2 * diff * d_eigmat_dp[idx, idx, :]
+            elif ref.kind == "eig_offdiagonal":
+                row, col = ref.atom_indices[:2]
+                calc_value = float((qm_evecs.T @ hess @ qm_evecs)[row, col])
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/evaluators/eigenmatrix.py]: Fixed in 8b9e7e9. Added bounds checks for both diagonal and off-diagonal refs, and hoisted eigmat computation out of the loop (addresses the next thread too).

File: q2mm/optimizers/evaluators/eigenmatrix.py [hunk 3]
```diff
@@ -161,17 +161,62 @@ def gradient(
...
+                diff = ref.value - calc_value
+                grad += -2.0 * ref.weight**2 * diff * d_eigmat_dp[row, col, :]
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/evaluators/eigenmatrix.py:238]: Fixed in 8b9e7e9. Hoisted `eigmat = qm_evecs.T @ hess @ qm_evecs` out of the loop — computed once and reused for all refs.

File: q2mm/optimizers/evaluators/hessian_element.py [hunk 4]
```diff
@@ -127,14 +127,40 @@ def gradient(
...
+        for ref in references:
+            if ref.atom_indices is None or len(ref.atom_indices) < 2:
+                raise ValueError(
+                    f"hessian_element requires atom_indices=(row, col), got {ref.atom_indices}. Label: {ref.label!r}"
+                )
+            row, col = ref.atom_indices[:2]
+            calc_value = float(hess[row, col])
+            diff = ref.value - calc_value
+            grad += -2.0 * ref.weight**2 * diff * dH_dp[row, col, :]
+        return grad
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/evaluators/hessian_element.py:174]: Fixed in 8b9e7e9. Added bounds check on (row, col) with label in error message, matching the pattern from `_extract()`.

File: q2mm/models/hessian.py [hunk 5]
```diff
@@ -266,6 +266,89 @@ def hessian_to_frequencies(
...
+
+    # Mass-weight the Hessian
+    hess = hessian_au.copy()
+    hess *= scale
+    hess = 0.5 * (hess + hess.T)
+
+    # Mass-weight each dH/dp slice
+    mw_dH_dp = np.empty((n3, n3, n_params))
+    for j in range(n_params):
+        mw_dH_dp[:, :, j] = dH_dp_au[:, :, j] * scale
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/models/hessian.py:324]: Fixed in 8b9e7e9. Added ndim and shape validation for both `dH_dp_au` and `hessian_au` at the top of `frequency_param_jacobian()`.

File: q2mm/optimizers/evaluators/frequency.py [hunk 6]
```diff
@@ -115,17 +115,45 @@ def gradient(
...
-            ``None`` — analytical gradients are not yet supported.
+            Gradient vector of shape ``(n_params,)``.
 
         """
-        return None
+        from q2mm.models.hessian import frequency_param_jacobian
+
+        target = structure if structure is not None else mol
+        hess, dH_dp = engine.hessian_and_param_jacobian(target, ff)
+        freqs, d_freq_dp = frequency_param_jacobian(hess, dH_dp, mol.symbols)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/evaluators/frequency.py:149]: Fixed in 11f1843. Added TypeError guard matching the EnergyEvaluator pattern. Skipped the d_freq_dp shape check since frequency_param_jacobian already validates shapes internally (added in 8b9e7e9).

File: q2mm/optimizers/evaluators/hessian_element.py [hunk 7]
```diff
@@ -127,14 +127,45 @@ def gradient(
...
 
         Returns:
-            ``None`` — analytical gradients are not yet supported.
+            Gradient vector of shape ``(n_params,)``.
 
         """
-        return None
+        target = structure if structure is not None else mol
+        hess, dH_dp = engine.hessian_and_param_jacobian(target, ff)
+
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/evaluators/hessian_element.py:158]: Fixed in 11f1843. Added TypeError guard with engine name in the message, same pattern as EnergyEvaluator.

File: q2mm/optimizers/evaluators/eigenmatrix.py [hunk 8]
```diff
@@ -161,17 +161,78 @@ def gradient(
...
         Returns:
-            ``None`` — analytical gradients are not yet supported.
+            Gradient vector of shape ``(n_params,)``.
 
         """
-        return None
+        from q2mm.models.hessian import decompose
+
+        target = structure if structure is not None else mol
+        hess, dH_dp = engine.hessian_and_param_jacobian(target, ff)
```
[the remark this answers is not in the document]
Comment (ericchansen) [reply q2mm/optimizers/evaluators/eigenmatrix.py:193]: Fixed in 11f1843. Added TypeError guard. Skipped the dH_dp shape assertion since the shape is already validated upstream in frequency_param_jacobian (8b9e7e9).
